pubsub/trace.go

```diff
@@ -256,3 +265,84 @@ func withSubscriptionKey(ctx context.Context, subName string) context.Context {
 func recordStat(ctx context.Context, m *stats.Int64Measure, n int64) {
 	stats.Record(ctx, m.M(n))
 }
+
+const defaultTracerName = "cloud.google.com/go/pubsub"
+
+func tracer() trace.Tracer {
+	return otel.Tracer(defaultTracerName, trace.WithInstrumentationVersion(internal.Version))
+}
+
+var _ propagation.TextMapCarrier = (*messageCarrier)(nil)
+
+// messageCarrier injects and extracts traces from a pubsub.Message.
+type messageCarrier struct {
+	msg *Message
+}
+
+// newMessageCarrier creates a new PubsubMessageCarrier.
+func newMessageCarrier(msg *Message) messageCarrier {
+	return messageCarrier{msg: msg}
+}
+
+// Get retrieves a single value for a given key.
+func (c messageCarrier) Get(key string) string {
+	return c.msg.Attributes["googclient_"+key]
+}
+
+// Set sets an attribute.
+func (c messageCarrier) Set(key, val string) {
+	c.msg.Attributes["googclient_"+key] = val
+}
+
+// Keys returns a slice of all keys in the carrier.
+func (c messageCarrier) Keys() []string {
+	i := 0
+	out := make([]string, len(c.msg.Attributes))
+	for k := range c.msg.Attributes {
+		out[i] = k
+		i++
+	}
+	return out
+}
+
+const (
+	// span names
+	publisherSpanName          = "send"
+	publishFlowControlSpanName = "publisher flow control"
+	publishSchedulerSpanName   = "publish scheduler"
+	publishRPCSpanName         = "send Publish"
```

Comment:
The spec doc seems to imply that this should be just "publish"?

Reply:
Yeah this should just be Publish, I went back and forth on this for a bit.